src/template.rs

```diff
@@ -146,81 +121,169 @@ impl Template {
+    /// A reference to the source event, to be copied from the relevant field or tag.
+    Reference(String),
+}
+
+// Wrap the parsed time formatter in order to provide `impl Hash` and some convenience functions.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+struct ParsedStrftime(Box<[Item<'static>]>);
+
+impl ParsedStrftime {
+    fn parse(fmt: &str) -> Result<Self, TemplateParseError> {
+        Ok(Self(
+            StrftimeItems::new(fmt)
+                .map(|item| match item {
+                    // Box the references so they outlive the reference
+                    Item::Space(space) => Item::OwnedSpace(space.into()),
+                    Item::Literal(lit) => Item::OwnedLiteral(lit.into()),
+                    // And copy all the others
+                    Item::Fixed(f) => Item::Fixed(f),
+                    Item::Numeric(num, pad) => Item::Numeric(num, pad),
+                    Item::Error => Item::Error,
+                    Item::OwnedSpace(space) => Item::OwnedSpace(space),
+                    Item::OwnedLiteral(lit) => Item::OwnedLiteral(lit),
+                })
+                .map(|item| {
+                    matches!(item, Item::Error)
+                        .then(|| Err(TemplateParseError::StrftimeError))
+                        .unwrap_or(Ok(item))
+                })
+                .collect::<Result<Vec<_>, _>>()?
+                .into(),
+        ))
+    }
+
+    fn is_dynamic(&self) -> bool {
+        self.0.iter().any(|item| match item {
+            Item::Fixed(_) => true,
+            Item::Numeric(_, _) => true,
+            Item::Error
+            | Item::Space(_)
+            | Item::OwnedSpace(_)
+            | Item::Literal(_)
+            | Item::OwnedLiteral(_) => false,
+        })
+    }
+
+    fn as_items(&self) -> impl Iterator<Item = &Item<'static>> + Clone {
+        self.0.iter()
     }
 }
 
-fn render_fields(src: &str, event: EventRef<'_>) -> Result<String, TemplateRenderingError> {
+fn parse_literal(src: &str) -> Result<Part, TemplateParseError> {
+    let parsed = ParsedStrftime::parse(src)?;
+    Ok(if parsed.is_dynamic() {
+        Part::Strftime(parsed)
+    } else {
+        Part::Literal(src.to_string())
+    })
+}
+
+// Pre-parse the template string into a series of parts to be filled in at render time.
+fn parse_template(src: &str) -> Result<Vec<Part>, TemplateParseError> {
+    let mut last_end = 0;
+    let mut parts = Vec::new();
+    for cap in RE.captures_iter(src) {
+        let all = cap.get(0).expect("Capture 0 is always defined");
+        if all.start() > last_end {
+            parts.push(parse_literal(&src[last_end..all.start()])?);
+        }
+        parts.push(Part::Reference(cap[1].trim().to_owned()));
+        last_end = all.end();
+    }
+    if src.len() > last_end {
+        parts.push(parse_literal(&src[last_end..])?);
+    }
+
+    Ok(parts)
+}
+
+fn render_fields(parts: &[Part], event: EventRef<'_>) -> Result<String, TemplateRenderingError> {
```

Comment:
maybe rename to `render_parts`? ("fields" referred to just the field lookups previously)